File: Instruction.cpp
#include "Instruction.h"

using namespace std;

extern vector<string> usedValues;			// Defined in Main.cpp

/**
 * Constructor
 */
Instruction::Instruction(string& instruction, vector<string>& usedValues)
{
	string opcode = instruction.substr(0, instruction.find(" "));
	string operands = instruction.substr(opcode.size(), instruction.size());
	string gram = opcode;
	string gramSimple = opcode;
	vector<string> listOfOperands;

	// Cycle through opperands if they exist
	if (operands != "") {
		stringstream opStream(operands);

		while (opStream.good()) {
			string operand;

			getline(opStream, operand, ',');
			operand = trim(operand);
			listOfOperands.push_back(operand);

			// Check for arithmetic in operand
			if (operand.find('+') != string::npos) {
				stringstream expStream(operand);
				string expressionValue;
				string expressionOperandType;
				int expressionVariable;

				int counter = 0;
				while (expStream.good()) {
					getline(expStream, expressionValue, '+');
					expressionValue = trim(expressionValue);

					expressionVariable = getVariable(operand, usedValues);
					this->variables.push_back(expressionVariable);

					expressionOperandType = getOperandType(operand);


					// Update gram representation
					if (counter <= 0) {
						gram += "." + expressionOperandType + to_string(expressionVariable);
					} else {
						gram += "+" + expressionOperandType + to_string(expressionVariable);
					}
					gramSimple += "." + expressionOperandType;

					counter++;
				}
			// No arithmetic found
			} else {
				int variable = getVariable(operand, usedValues);
				string operandType = getOperandType(operand);

				this->variables.push_back(variable);

				gram += "." + operandType + to_string(variable);
				gramSimple += "." + operandType;
			}
		}
	}

	// Set values for new instruction
	this->opCode = opcode;
	this->gram = gram;
	this->gramSimple = gramSimple;
	this->listOfOperands = listOfOperands;
}

/**
 * Checks if variable exists, if not start tracking it
 */
int Instruction::getVariable(string& operand, vector<string>& usedValues)
{
	int position = find(usedValues.begin(), usedValues.end(), operand) - usedValues.begin();

	if (position < usedValues.size()) {
		return position;
	}

	usedValues.push_back(operand);

	return usedValues.size() - 1;
}

/**
 * Removes unwanted chars on str
 */
string Instruction::trim(string& str)
{
	size_t start = str.find_first_not_of(" ");
	size_t end = str.find_last_not_of(" ");

	str.erase(remove(str.begin(), str.end(), '\t'), str.end());
	str.erase(remove(str.begin(), str.end(), '\r'), str.end());
	str.erase(remove(str.begin(), str.end(), '\n'), str.end());

	return str.substr(start, end - start + 1);
}

/**
 * Gets the type of operand for gram
 */
string Instruction::getOperandType(string& operand)
{
	if (operand.substr(0, 2) == "#0") {
		return "con";
	} else if (operand.substr(0, 2) == "0x") {
		return "mem";
	}

	return "reg";
}

/**
 * Print variables for operands
 */
void Instruction::printVariables()
{
	for (int i = 0; i < this->variables.size(); i++) {
		cout << usedValues[this->variables[i]] << endl;
	}
}

File: Instruction.py
import Global

class Instruction:
    """
    self.text: original instruction text (Ex: lda al 0x1070)
    self.mnemonic: operation done in instruction
    self.operands: array of instruction operands
    self.gram: mnemonic.operands with operand type (Ex: lda.reg.mem)
    """

    def __init__(self, file_line):
        """
        Instruction constructor
        :param file_line: array of split instruction line
        """
        self.text = ' '.join(file_line)
        self.mnemonic = file_line[0]
        self.operands = file_line[1:]
        self.gram = self.mnemonic

        for i in range(len(file_line[1:])):
            operand = file_line[i + 1]

            # Handle operand arithmetic
            if operand == "+":
                self.gram += "+"
            else:
                if file_line[i] == "+":
                    self.gram += Global.get_operand_type(operand)
                else:
                    self.gram += "." + Global.get_operand_type(operand)

    def __str__(self):
        """
        :return: formatted string with all instance variables
        """
        return ("Text :     " + self.text + "\n" +
                "Mnemonic : " + self.mnemonic + "\n" +
                "Gram :     " + self.gram + "\n")
